File: client/old/src/main/java/net/treset/minecraftlauncher/sync/ComponentList.java
package net.treset.minecraftlauncher.sync;

import net.treset.mc_version_loader.json.GenericJsonParsable;
import net.treset.mc_version_loader.json.SerializationException;
import net.treset.minecraftlauncher.LauncherApplication;

import java.util.Arrays;
import java.util.List;

public class ComponentList extends GenericJsonParsable {
    public static class Entry {
        private final String id;
        private final String name;

        public Entry(String id, String name) {
            this.id = id;
            this.name = name;
        }

        public String getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        @Override
        public String toString() {
            return (name.isBlank() ? LauncherApplication.stringLocalizer.get("sync.component.unknown") : name) + " (" + (id.length() > 8 ? id.substring(0, 7) : id) + "...)";
        }
    }

    public static ComponentList fromJson(String json) throws SerializationException {
        Entry[] entries = GenericJsonParsable.fromJson(json, Entry[].class);
        return new ComponentList(Arrays.asList(entries));
    }

    private final List<Entry> entries;

    public ComponentList(List<Entry> entries) {
        this.entries = entries;
    }

    public List<Entry> getEntries() {
        return entries;
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/sync/ComponentList.kt
package net.treset.treelauncher.backend.sync

import net.treset.mcdl.json.GenericJsonParsable
import net.treset.mcdl.json.SerializationException
import net.treset.treelauncher.localization.strings
import java.util.*

class ComponentList(val entries: List<Entry>) : GenericJsonParsable() {
    class Entry(val id: String, val name: String) {

        override fun toString(): String {
            return "${name.ifBlank { strings().sync.unknown() }} (${if (id.length > 8) id.substring(0, 7) else "$id..."})"
        }
    }

    companion object {
        @Throws(SerializationException::class)
        fun fromJson(json: String?): ComponentList {
            val entries = fromJson(json, Array<Entry>::class.java)
            return ComponentList(listOf(*entries))
        }
    }
}
